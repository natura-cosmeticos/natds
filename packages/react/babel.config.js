const path = require('path')

const defaultAlias = {
  '@naturacosmeticos/natds-react': path.resolve(__dirname, './src/components')
}

module.exports = {
  plugins: [
    ['babel-plugin-module-resolver', { alias: defaultAlias }]
  ],
  presets: [
    ['@babel/preset-env', { targets: { node: 'current' } }],
    '@babel/preset-typescript',
    '@babel/preset-react'
  ]
}
