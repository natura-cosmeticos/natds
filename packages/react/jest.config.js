const base = require('../../jest.config.base.js')
const { displayName, name } = require('./package.json')

module.exports = {
  ...base,
  collectCoverageFrom: [
    'src/**/*.{js,jsx,ts,tsx}',
    '!src/**/*.stories.tsx',
    '!src/helpers/**'
  ],
  errorOnDeprecated: true,
  displayName,
  name,
  roots: ['<rootDir>/src'],
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest',
    '^.+\\.mdx?$': '@storybook/addon-docs/jest-transform-mdx'
  }
}
