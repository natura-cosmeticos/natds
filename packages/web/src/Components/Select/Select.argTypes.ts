import { ArgType } from '@storybook/addons'
import { ISelectProps } from './Select.props'
import { booleanArgType } from '../../../.storybook/argTypes/booleanArgType'
import { textArgType } from '../../../.storybook/argTypes/textArgType'

export const argTypes: Record<keyof ISelectProps, ArgType> = {
  disabled: {
    ...booleanArgType,
    description: 'Disables the `select` for user interactions.',
    table: {
      ...booleanArgType.table,
      defaultValue: { summary: false }
    }
  },
  helpText: {
    ...textArgType,
    description: 'The text below the `select` element.'
  },
  id: {
    ...textArgType,
    description:
      'The `id` of the wrapper element or the `select` element when `native`.'
  },
  label: {
    ...textArgType,
    description:
      'Label is the text used above the Select element. It describes the field name.'
  },
  onChange: {
    defaultValue: null,
    type: 'function'
  },
  options: {
    control: { type: 'array' },
    description: 'All options of Select',
    table: {
      defaultValue: { summary: '[one, two, three]' },
      type: { summary: 'array' }
    }
  },
  placeholder: {
    ...textArgType,
    description: ''
  },
  required: {
    ...booleanArgType,
    description: 'Whether the select is required or not.',
    table: {
      ...booleanArgType.table,
      defaultValue: { summary: false }
    }
  },
  state: {
    control: { options: ['success', 'error'], type: 'inline-radio' },
    description: 'Sets whether the `input` is in `error` or `success` state.',
    table: { type: { summary: '"success" | "error" | undefined' } }
  }
}

export default argTypes
