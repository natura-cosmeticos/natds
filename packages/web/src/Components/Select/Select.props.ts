import { State } from 'Components/Input/Input.props';

export interface SelectProps {
  /**
   * @optional
   * Disables the Select interaction
   */
  disabled?: boolean;

  /**
   * @optional
   * Help text placed underneath the Select
   */
  helpText?: string;

  /**
   * The Select id property
   */
  id?: string;

  /**
   * @optional
   * Help text placed underneath the Select
   */
  label?: string;

  /**
   * Options of Select
   */
  options: string[];

  /**
   * @optional
   */
  placeholder: string;

  /**
   * @optional
   * If this Select is required or not
   */
  required?: boolean;
  /**
   * @optional
   * Optional state for component variant
   */
  state?: State;
}

export type ISelectProps = SelectProps;
