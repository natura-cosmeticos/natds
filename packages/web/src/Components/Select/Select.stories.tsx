/* eslint-disable complexity */
import { Meta, Story } from '@storybook/react'
import * as React from 'react'
import { Select as component } from './Select'
import { ISelectProps } from './Select.props'

const Select: React.ForwardRefExoticComponent<ISelectProps> = React.lazy(() => import('./Select'))

export default {
  component,
  title: 'Components/Select'
} as Meta

const options = ['option 1', 'option 2', 'option 3']

const Template: Story<ISelectProps> = (args: ISelectProps) => {
  const [value, setValue] = React.useState(args.value)

  const handleChange = (event: React.ChangeEvent<{ value: unknown }>) => {
    setValue(event.target.value as string)
  }

  return (
    <Select
      onChange={handleChange}
      value={args.value || value}
      options={args.options || options}
      label={args.label || 'Label'}
      helpText={args.helpText || 'Assistive text'}
      id={args.id || 'demo-select'}
      placeholder={args.placeholder || 'Placeholder'}
      {...args}
    />
  )
}

export const Playground: Story<ISelectProps> = Template.bind({})
Playground.args = {
  disabled: false,
  helpText: 'Helper Text',
  label: 'Label',
  required: false,
  options,
  placeholder: 'Placeholder'
}
