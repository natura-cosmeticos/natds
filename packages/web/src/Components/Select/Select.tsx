import * as React from 'react'
import MaterialSelect from '@material-ui/core/Select'
import { MenuProps } from '@material-ui/core'
import { Icon } from '../Icon'
import { MenuItem } from '../MenuItem'
import InputStateHelpTextProvider from '../InputStateHelpTextProvider'
import { ISelectProps } from './Select.props'

export { ISelectProps } from './Select.props'

/**
 * ## Importing
 *
 * ```
 * import { Select } from '@naturacosmeticos/natds-web';
 * ```
 */

export const Select = React.forwardRef<HTMLSelectElement | HTMLInputElement, ISelectProps>(
  (props: ISelectProps, ref) => {
    const {
      defaultValue,
      id,
      multiple,
      onChange,
      onClose,
      onOpen,
      options,
      placeholder,
      state,
      value,
      ...rest
    } = props

    const menuProps: Partial<MenuProps> = {
      anchorOrigin: {
        vertical: 'bottom',
        horizontal: 'left'
      },
      transformOrigin: {
        vertical: 'top',
        horizontal: 'left'
      },
      getContentAnchorEl: null
    }

    return (
      <InputStateHelpTextProvider {...rest} state={state}>
        <MaterialSelect
          {...props}
          data-state={state}
          defaultValue={defaultValue}
          displayEmpty
          IconComponent={(props) => (<Icon name="outlined-navigation-arrowbottom" {...props} />)}
          id={id}
          MenuProps={menuProps}
          multiple={multiple}
          onChange={onChange}
          onClose={onClose}
          onOpen={onOpen}
          ref={ref}
          value={value}
        >
          {placeholder && (
            <MenuItem value="" disabled>
              {placeholder}
            </MenuItem>
          )}
          {options?.map((option, index) => (
            <MenuItem value={option} key={index}>
              {option}
            </MenuItem>
          ))}
        </MaterialSelect>
      </InputStateHelpTextProvider>
    )
  }
)

Select.displayName = 'Select'

export default Select
