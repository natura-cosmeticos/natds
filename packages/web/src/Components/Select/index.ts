import { Select } from './Select'

export { ISelectProps } from './Select.props'
export default Select
