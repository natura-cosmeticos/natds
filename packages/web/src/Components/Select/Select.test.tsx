import * as React from 'react'
import * as TestRenderer from 'react-test-renderer'
import { render, fireEvent, within } from '@testing-library/react'
import { ISelectProps, Select, handleOptions } from './Select'

const getProps = (props: Partial<ISelectProps> = {}) => {
  const {
    disabled,
    helpText,
    id = 'Select',
    label,
    options = ['Option1', 'Option2'],
    placeholder = 'Placeholder',
    state,
    onChange
  } = props

  return {
    state,
    id,
    placeholder,
    options,
    disabled,
    helpText,
    label,
    onChange
  }
}

describe('Select component', () => {
  it('should render correctly', () => {
    const component = TestRenderer.create(<Select {...getProps()} />)
    expect(component).toMatchSnapshot()
  })
  it('should render as disabled', () => {
    const component = TestRenderer.create(
      <Select {...getProps()} disabled />
    )
    expect(component).toMatchSnapshot()
  })
  it('should render with helper text', () => {
    const component = TestRenderer.create(
      <Select {...getProps()} helpText="Helper Text" />
    )
    expect(component).toMatchSnapshot()
  })
  it('should render with label', () => {
    const component = TestRenderer.create(
      <Select {...getProps()} label="Label" />
    )
    expect(component).toMatchSnapshot()
  })
  it('should render as required', () => {
    const component = TestRenderer.create(
      <Select {...getProps()} required />
    )
    expect(component).toMatchSnapshot()
  })
  it('should render with state success', () => {
    const component = TestRenderer.create(
      <Select {...getProps()} state="success" />
    )
    expect(component).toMatchSnapshot()
  })
  it('should render with state error', () => {
    const component = TestRenderer.create(
      <Select {...getProps()} state="error" />
    )
    expect(component).toMatchSnapshot()
  })
  it('should select option1', async () => {
    const { getByRole } = render(<Select {...getProps()} />)

    fireEvent.mouseDown(getByRole('button'))
    const listbox = within(getByRole('listbox'))
    fireEvent.click(listbox.getByText('Option1'))

    expect(getByRole('button')).toMatchSnapshot()
  })
  it('should select option2', async () => {
    const { getByRole } = render(<Select {...getProps()} />)

    fireEvent.mouseDown(getByRole('button'))
    const listbox = within(getByRole('listbox'))
    fireEvent.click(listbox.getByText('Option2'))

    expect(getByRole('button')).toMatchSnapshot()
  })
  it('should calls onChange when select option1', async () => {
    const mockOnChange = jest.fn()
    const { getByRole } = render(
      <Select {...getProps()} onChange={mockOnChange} />
    )

    fireEvent.mouseDown(getByRole('button'))

    const listbox = within(getByRole('listbox'))

    fireEvent.click(listbox.getByText('Option1'))

    expect(mockOnChange).toHaveBeenCalledTimes(1)
    expect(getByRole('button').textContent).toEqual('Option1')
  })

  describe('handleOptions', () => {
    it('should return correctly when option is number', () => {
      expect(handleOptions([1, 2, 3])).toEqual([
        { value: 1, description: 1 },
        { value: 2, description: 2 },
        { value: 3, description: 3 }
      ])
    })
    it('should return correctly when option is string', () => {
      expect(handleOptions(['1', '2', '3'])).toEqual([
        { value: '1', description: '1' },
        { value: '2', description: '2' },
        { value: '3', description: '3' }
      ])
    })
    it('should return correctly when option is object', () => {
      expect(handleOptions([
        { value: 1, description: 'one' },
        { value: 2, description: 'two' },
        { value: 3, description: 'three' }
      ])).toEqual([
        { value: 1, description: 'one' },
        { value: 2, description: 'two' },
        { value: 3, description: 'three' }
      ])
    })
  })
})
