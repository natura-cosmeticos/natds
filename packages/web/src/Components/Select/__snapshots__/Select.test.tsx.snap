// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component should render as disabled 1`] = `
<div
  className="MuiFormControl-root NatDSFormControl-root-1"
>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline Mui-disabled Mui-disabled MuiInputBase-formControl MuiInput-formControl"
    onClick={[Function]}
    options={
      Array [
        "Option1",
        "Option2",
      ]
    }
  >
    <div
      aria-disabled="true"
      aria-haspopup="listbox"
      aria-labelledby="Select"
      className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input Mui-disabled Mui-disabled Mui-disabled"
      id="Select"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={null}
      role="button"
      tabIndex={null}
    >
      <span
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#8203;",
          }
        }
      />
    </div>
    <input
      aria-hidden={true}
      className="MuiSelect-nativeInput"
      onAnimationStart={[Function]}
      onChange={[Function]}
      placeholder="Placeholder"
      required={false}
      tabIndex={-1}
    />
    <i
      aria-hidden={true}
      className="material-icons MuiIcon-root natds-icons natds-icons-outlined-navigation-arrowbottom MuiSelect-icon Mui-disabled NatDSIcon-root-2 NatDSIcon-root-4"
    />
  </div>
</div>
`;

exports[`Select component should render as required 1`] = `
<div
  className="MuiFormControl-root NatDSFormControl-root-1"
>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    onClick={[Function]}
    options={
      Array [
        "Option1",
        "Option2",
      ]
    }
    required={true}
  >
    <div
      aria-haspopup="listbox"
      aria-labelledby="Select"
      className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
      id="Select"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#8203;",
          }
        }
      />
    </div>
    <input
      aria-hidden={true}
      className="MuiSelect-nativeInput"
      onAnimationStart={[Function]}
      onChange={[Function]}
      placeholder="Placeholder"
      required={true}
      tabIndex={-1}
    />
    <i
      aria-hidden={true}
      className="material-icons MuiIcon-root natds-icons natds-icons-outlined-navigation-arrowbottom MuiSelect-icon NatDSIcon-root-2 NatDSIcon-root-12"
    />
  </div>
</div>
`;

exports[`Select component should render correctly 1`] = `
<div
  className="MuiFormControl-root NatDSFormControl-root-1"
>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    onClick={[Function]}
    options={
      Array [
        "Option1",
        "Option2",
      ]
    }
  >
    <div
      aria-haspopup="listbox"
      aria-labelledby="Select"
      className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
      id="Select"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#8203;",
          }
        }
      />
    </div>
    <input
      aria-hidden={true}
      className="MuiSelect-nativeInput"
      onAnimationStart={[Function]}
      onChange={[Function]}
      placeholder="Placeholder"
      required={false}
      tabIndex={-1}
    />
    <i
      aria-hidden={true}
      className="material-icons MuiIcon-root natds-icons natds-icons-outlined-navigation-arrowbottom MuiSelect-icon NatDSIcon-root-2 NatDSIcon-root-3"
    />
  </div>
</div>
`;

exports[`Select component should render with helper text 1`] = `
<div
  className="MuiFormControl-root NatDSFormControl-root-1"
>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    helpText="Helper Text"
    onClick={[Function]}
    options={
      Array [
        "Option1",
        "Option2",
      ]
    }
  >
    <div
      aria-haspopup="listbox"
      aria-labelledby="Select"
      className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
      id="Select"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#8203;",
          }
        }
      />
    </div>
    <input
      aria-hidden={true}
      className="MuiSelect-nativeInput"
      onAnimationStart={[Function]}
      onChange={[Function]}
      placeholder="Placeholder"
      required={false}
      tabIndex={-1}
    />
    <i
      aria-hidden={true}
      className="material-icons MuiIcon-root natds-icons natds-icons-outlined-navigation-arrowbottom MuiSelect-icon NatDSIcon-root-2 NatDSIcon-root-5"
    />
  </div>
  <p
    className="MuiFormHelperText-root NatDSFormHelperText-root-6 NatDSFormHelperText-root-8"
  >
    Helper Text
  </p>
</div>
`;

exports[`Select component should render with label 1`] = `
<div
  className="MuiFormControl-root NatDSFormControl-root-1"
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root NatDSInputLabel-root-9 NatDSInputLabel-root-10 MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
    data-shrink={true}
  >
    Label
  </label>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    onClick={[Function]}
    options={
      Array [
        "Option1",
        "Option2",
      ]
    }
  >
    <div
      aria-haspopup="listbox"
      aria-labelledby="Select"
      className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
      id="Select"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#8203;",
          }
        }
      />
    </div>
    <input
      aria-hidden={true}
      className="MuiSelect-nativeInput"
      onAnimationStart={[Function]}
      onChange={[Function]}
      placeholder="Placeholder"
      required={false}
      tabIndex={-1}
    />
    <i
      aria-hidden={true}
      className="material-icons MuiIcon-root natds-icons natds-icons-outlined-navigation-arrowbottom MuiSelect-icon NatDSIcon-root-2 NatDSIcon-root-11"
    />
  </div>
</div>
`;

exports[`Select component should render with state error 1`] = `
<div
  className="MuiFormControl-root NatDSFormControl-root-1"
>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    data-state="error"
    onClick={[Function]}
    options={
      Array [
        "Option1",
        "Option2",
      ]
    }
    state="error"
  >
    <div
      aria-haspopup="listbox"
      aria-labelledby="Select"
      className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
      id="Select"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#8203;",
          }
        }
      />
    </div>
    <input
      aria-hidden={true}
      className="MuiSelect-nativeInput"
      onAnimationStart={[Function]}
      onChange={[Function]}
      placeholder="Placeholder"
      required={false}
      tabIndex={-1}
    />
    <i
      aria-hidden={true}
      className="material-icons MuiIcon-root natds-icons natds-icons-outlined-navigation-arrowbottom MuiSelect-icon NatDSIcon-root-2 NatDSIcon-root-14"
    />
  </div>
</div>
`;

exports[`Select component should render with state success 1`] = `
<div
  className="MuiFormControl-root NatDSFormControl-root-1"
>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    data-state="success"
    onClick={[Function]}
    options={
      Array [
        "Option1",
        "Option2",
      ]
    }
    state="success"
  >
    <div
      aria-haspopup="listbox"
      aria-labelledby="Select"
      className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
      id="Select"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#8203;",
          }
        }
      />
    </div>
    <input
      aria-hidden={true}
      className="MuiSelect-nativeInput"
      onAnimationStart={[Function]}
      onChange={[Function]}
      placeholder="Placeholder"
      required={false}
      tabIndex={-1}
    />
    <i
      aria-hidden={true}
      className="material-icons MuiIcon-root natds-icons natds-icons-outlined-navigation-arrowbottom MuiSelect-icon NatDSIcon-root-2 NatDSIcon-root-13"
    />
  </div>
</div>
`;

exports[`Select component should select option1 1`] = `
<div
  aria-haspopup="listbox"
  aria-labelledby="Select"
  class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
  id="Select"
  role="button"
  tabindex="0"
>
  Option1
</div>
`;

exports[`Select component should select option2 1`] = `
<div
  aria-haspopup="listbox"
  aria-labelledby="Select"
  class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
  id="Select"
  role="button"
  tabindex="0"
>
  Option2
</div>
`;
